### PR TITLE
[FIX] sms: don't get error when sending an sms on sale order

### PR DESCRIPTION
**Pull request by mamatch**
steps:
- Go to Rental app
- Open a rental Order
- Click on Action
-  Send an SMS text message
-  Fill phone number and message
- Sens SMS 

Issue:
Traceback

Cause:
Sale order doesn't have a phone or mobile field so the sms.composer tries to write on it use "False"

Solution ensure the field name isn't false before writing

opw-3103232

### CODE REVIEW

File: addons/sms/tests/test_sms_template.py [hunk 1]
```diff
@@ -106,3 +106,28 @@ def test_sms_template_rendering_unrestricted(self):
...
+            'number_field_name': False,
+            'numbers': False,
+            'recipient_single_number_itf': '+32494670905',
+            'res_id': sale_order.id,
+            'res_model': 'sale.order'
+        })
+        try:
+            sms_composer.action_send_sms()
+        except ValueError as e:
+            self.fail(f"{e}\nThe action must be done despite the empty `number_field_name` field")
```
Comment (thiryr) [inline addons/sms/tests/test_sms_template.py:133]: Maybe check the message is sent while we're at it, this way the test is more generally applicable.
(It's easy to check, you can look at all the other tests in that file)

File: addons/sms/tests/test_sms_template.py [hunk 2]
```diff
@@ -106,3 +106,28 @@ def test_sms_template_rendering_unrestricted(self):
 
         body = sms_composer._prepare_body_values(self.partner)[self.partner.id]
         self.assertIn(self.partner.name, body, 'Template Editor should be able to write new Jinja code')
+
+@tagged("post_install")
+class TestSMSComposer(TransactionCase):
```
Comment (thiryr) [inline addons/sms/tests/test_sms_template.py:111]: We have addons specifically for mail/sms tests. This one should go in ``test_mail_sms``, in ``test_sms_composer.py``

**Review by reth-odoo — COMMENTED** [review]
Hello :)

Small remarks for the test and then I think it'll be good for me

File: addons/sms/tests/test_sms_template.py (hunk 2 above)
Comment (mamatch) [reply addons/sms/tests/test_sms_template.py:111]: Hi, I tried this in test_sms_composer.py but I get an error, I don't know why
```python
    def test_composer_sending_on_sale_order(self):
        partner = self.env['res.partner'].create({'name': 'Test Partner'})
        sale_order = self.env['sale.order'].create({
            'partner_id': partner.id,
        })

        sms_composer = self.env['sms.composer'].create({
            'body': self._test_body,
            'composition_mode': 'comment',
            'mass_force_send': False,
            'mass_keep_log': True,
            'number_field_name': False,
            'numbers': False,
            'recipient_single_number_itf': self.random_numbers_san[0],
            'res_id': sale_order.id,
            'res_model': 'sale.order'
        })

        with self.mockSMSGateway():
            sms_composer._action_send_sms()

        self.assertSMSNotification([{'partner': partner, 'number': self.random_numbers_san[0]}], self._test_body)

``` 
Screenshot from 2023-01-09 15-53-10

Can you help me ? :)

File: addons/sms/wizard/sms_composer.py [hunk 3]
```diff
@@ -203,7 +203,7 @@ def _action_send_sms_comment_single(self, records=None):
         # on the numbers in the database.
         records = records if records is not None else self._get_records()
         records.ensure_one()
-        if self.recipient_single_number_itf and self.recipient_single_number_itf != self.recipient_single_number:
+        if self.recipient_single_number_itf and self.recipient_single_number_itf != self.recipient_single_number and self.number_field_name:
```
Comment (thiryr) [inline addons/sms/wizard/sms_composer.py:206]: Apparently we assume there should be 'numbers' or a number field in ``_action_send_sms_comment`` :)

I think just updating numbers should be fine here.

```suggestion
        if not self.number_field_name:
            self.numbers = self.recipient_single_number_itf or self.recipient_single_number
        elif self.recipient_single_number_itf and self.recipient_single_number_itf != self.recipient_single_number:
```

File: addons/sms/tests/test_sms_template.py (hunk 2 above)
Comment (thiryr) [reply addons/sms/tests/test_sms_template.py:111]: You should not use 'sale.order' as a test record, as it is not a dependency of this module. Use one of the test models available, tde pointed out ``mail.test.sms.partner`` was great for this.

You don't need to use a partner either, you can just do 
``test_record = self.env['mail.test.sms.partner'].create({'name': 'Test'})`` (update the composer too)
and 
``self.assertSMSNotification([{'number': self.random_numbers_san[0]}], self._test_body)``

**Review by reth-odoo — COMMENTED** [review]
Hello again :)

Seems like your test fails because of another issue, good thing we added a test :D

**Review by reth-odoo — APPROVED** [review]
Looking amazing to me, thanks for the thorough follow-up :)

**Comment by thiryr** [comment]
:skull: Guess there was no point moving this to 15.3 in the end (you can just retarget this and rebase, of course)

**Comment by mamatch** [comment]
Is it normal to have so much conflicts ? Can't we just re-open the initial branch on 16.0 and then push the changes ?

**Comment by thiryr** [comment]
@mamatch :/ yeah, you should probably do that.

You can retrieve your commit from the runbot, just go to your last valid build (/batch/123456) > commits > click the github icon. Here's this one 975491ddd317575ba31c754d6d812ff29e58387e

**Comment by mamatch** [comment]
Seems like I can't reopen :/
Screenshot from 2023-01-12 12-18-31

**Comment by thiryr** [comment]
Yeah don't worry PRs are cheap, just close this one and link it in a new one.
Then cherrypick your last commit 975491ddd317575ba31c754d6d812ff29e58387e (`975491ddd317575ba31c754d6d812ff29e58387e`)

**Comment by mamatch** [comment]
Here is it :) https://github.com/odoo/odoo/pull/109756

**Pull request closed without merging**